external/js/cothority/src/network/connection.ts

```diff
@@ -132,42 +131,85 @@ export class WebSocketConnection implements IConnection {
     addresses: string[];
+    addressNext: number;
+    connections: WebSocketConnection[] = [];
 
     /**
      * @param r         The roster to use
      * @param service   The name of the service to reach
-     * @param retry = 2 How many times to retry a failing connection with another node
      */
-    constructor(r: Roster, service: string, private retry: number = 2) {
-        super("", service);
+    constructor(r: Roster, service: string) {
         this.addresses = r.list.map((conode) => conode.getWebSocketAddress());
         shuffle(this.addresses);
+        this.addressNext = RosterWSConnection.parallel;
+        if (this.addressNext > this.addresses.length) {
+            this.addressNext = this.addresses.length;
+        }
+        for (let i = 0; i < this.addressNext; i++) {
+            this.connections.push(new WebSocketConnection(this.addresses[i], service));
+        }
     }
 
-    /** @inheritdoc */
+    /**
+     * Sends a message to conodes in parallel. As soon as one of the conodes returns
+     * success, the message is returned. If a conode returns an error (or times out),
+     * a next conode from this.addresses is contacted. If all conodes return an error,
+     * the promise is rejected.
+     *
+     * @param message the message to send
+     * @param reply the type of the message to return
+     */
     async send<T extends Message>(message: Message, reply: typeof Message): Promise<T> {
         const errors: string[] = [];
-        for (let i = 0; i < this.addresses.length; i++) {
-            this.url = this.addresses[0];
-            Log.lvl3("sending", message.constructor.name, "to address", this.url);
-
-            try {
-                // we need to await here to catch and try another conode
-                return await super.send(message, reply);
-            } catch (e) {
-                Log.lvl3(`failed to send on ${this.url} with error:`, e);
-                errors.push(e.message);
-                if (i > this.retry) {
-                    return Promise.reject(errors.join(" :: "));
-                }
-                Log.error("Error while sending - trying with next node");
-                this.addresses = [...this.addresses.slice(1), this.url];
-            }
-        }
+        let rotate = this.addresses.length - this.connections.length;
+
+        // Get the first reply - need to take care not to return a reject too soon, else
+        // all other promises will be ignored.
+        return Promise.race(this.connections.map((connection) => {
+            return new Promise<T>(async (resolve, reject) => {
+                do {
+                    try {
+                        const sub = await connection.send(message, reply);
+                        // Signal to other connections that have an error that they don't need
+                        // to retry.
+                        rotate = -1;
+                        resolve(sub as T);
+                    } catch (e) {
```

Comment:
>  If a request fails and there's no more rotation, it won't reject if another promise resolved as they won't have enough errors to reject.

That's by design and exactly the point: I don't want to know if any connection failed, I want to know if any connection succeeds. I only want to know about failure if _all_ connections fail. Else I only want success.

Also, I don't understand how the accepted solution in the SO link is acceptable: to me it looks like it would wait for all connections to be either resolved or rejected, while the OP (and this case) wants to have the first successful connection.

And, yes, there will be open connections when the `Promise.race` returns. This seems to be by design of `Promise.race`, and I don't see how I can change that.

Reply:
> 
> That's by design and exactly the point: I don't want to know if any connection failed, I want to know if any connection succeeds. I only want to know about failure if _all_ connections fail. Else I only want success.
That's the problem, I got that you want to achieve that but the fact is that some promise could not be fulfilled at all, meaning they will be leaking. It should always either resolve or reject.

> 
> Also, I don't understand how the accepted solution in the SO link is acceptable: to me it looks like it would wait for all connections to be either resolved or rejected, while the OP (and this case) wants to have the first successful connection.
Well that allows you to resolve or reject any promises by using the property of `Promise.all`. `Promise.race` returns either with rejection or resolution **but** `Promise.all` returns only if all have resolve **or** on the first failure. So you inverse that you can get the first promise that resolves or all the rejections.

> 
> And, yes, there will be open connections when the `Promise.race` returns. This seems to be by design of `Promise.race`, and I don't see how I can change that.
I know, it's not the problem.

Reply:
Thanks for clarifying the SO solution - I didn't understand that they reject on success... Between rejecting on success and ignoring the promise, I'd chose the latter, which seems to still get garbage collected:

https://stackoverflow.com/questions/36734900/what-happens-if-we-dont-resolve-or-reject-the-promise

But if you think this is not OK, I'll change it around and reject on success.

Reply:
Hm no you're right. I missed the point that it will be GCed because of the race. I got stuck by the `all` logic.